### PR TITLE
feat: Add background job management with detached execution

### LINKED ISSUE #68: feat: Add background job management with detached execution
## Summary

Add the ability to run `lxa` commands in the background with proper job management, including detaching from the terminal, log capture, and job control.

## Problem

Currently, `lxa` commands like `implement --loop` and `refine` run in the foreground and are long-running. This creates several issues:

1. **Terminal exit kills the process** - Closing the terminal or SSH disconnect terminates work
2. **No way to check status** - Cannot see what's running or completed
3. **No log management** - Output is lost if terminal scrollback fills up
4. **Cannot run multiple jobs** - Only one foreground process per terminal

### Background Flag with Job Names

```bash
# Start a background job with auto-generated name
lxa implement --loop --background
# Output: Started job implement-a3f2b1c, logs at ~/.lxa/jobs/implement-a3f2b1c.log

# Start with custom job name (hash appended for uniqueness)
lxa implement --loop --background --job-name overnight-run
# Output: Started job overnight-run-f8c1d2e, logs at ~/.lxa/jobs/overnight-run-f8c1d2e.log

# Works with any long-running command
lxa refine https://github.com/owner/repo/pull/42 --background --job-name pr42-refine
```

### Job Naming Convention

- **Format**: `{name}-{hash}` where hash is 7 characters
- **Default name**: Derived from command (e.g., `implement`, `refine`)
- **Custom name**: Via `--job-name` flag
- **Hash**: First 7 chars of UUID4 or similar, ensures uniqueness without state tracking

### Detached Execution

The background process must survive terminal exit:

- Use `setsid` or `nohup` semantics to detach from controlling terminal
- Redirect stdout/stderr to log file
- Store PID in job metadata for control
- Process continues running after parent shell exits

### Job Control Commands

```bash
# List all jobs (running and recent completed)
lxa job list
# ID                    Command                    Status   Started      Duration
# overnight-run-f8c1d2e implement --loop           running  2h ago       2h 15m
# pr42-refine-a1b2c3d   refine PR#42               done     1d ago       45m
# implement-x9y8z7w     implement --loop           failed   2d ago       12m

# View job status
lxa job status overnight-run-f8c1d2e
# Job: overnight-run-f8c1d2e
# Command: lxa implement --loop
# Status: running
# PID: 12345
# Started: 2026-04-13 15:30:00
# Log: ~/.lxa/jobs/overnight-run-f8c1d2e.log

# Tail logs (follows like tail -f)
lxa job logs overnight-run-f8c1d2e --follow

# View last N lines
lxa job logs overnight-run-f8c1d2e --lines 100

# Stop a running job gracefully (SIGTERM, then SIGKILL after timeout)
lxa job stop overnight-run-f8c1d2e

# Clean up old job logs/metadata
lxa job clean --older-than 7d
```

### Job Metadata Storage

Store job state in `~/.lxa/jobs/`:

```
~/.lxa/jobs/
├── overnight-run-f8c1d2e.json   # Job metadata
├── overnight-run-f8c1d2e.log    # Stdout/stderr
├── pr42-refine-a1b2c3d.json
├── pr42-refine-a1b2c3d.log
└── ...
```

**Metadata file structure:**
```json
{
  "id": "overnight-run-f8c1d2e",
  "command": ["implement", "--loop"],
  "cwd": "/path/to/repo",
  "pid": 12345,
  "status": "running",
  "started_at": "2026-04-13T15:30:00Z",
  "ended_at": null,
  "exit_code": null
}
```

### Detach Mechanism

On Unix-like systems:
1. Fork the process
2. Call `setsid()` to create new session (detach from terminal)
3. Redirect stdin from `/dev/null`
4. Redirect stdout/stderr to log file
5. Write PID to metadata file
6. Parent exits immediately after reporting job ID

### Status Detection

- Check if PID exists and is our process (via `/proc/{pid}/cmdline` or `ps`)
- Update metadata on completion (exit code, ended_at)
- Handle stale PIDs (process died without cleanup)

### Log Rotation (Future)

Not in initial scope, but consider:
- Max log file size
- Automatic rotation
- Compression of old logs

## Acceptance Criteria

- [ ] `--background` flag starts detached process that survives terminal exit
- [ ] `--job-name` allows custom naming with auto-appended hash
- [ ] `lxa job list` shows running and recent jobs
- [ ] `lxa job status <id>` shows detailed job info
- [ ] `lxa job logs <id>` displays log contents
- [ ] `lxa job logs <id> --follow` tails logs in real-time
- [ ] `lxa job stop <id>` gracefully terminates running jobs
- [ ] Jobs directory created at `~/.lxa/jobs/`
- [ ] Metadata persists across restarts

## Related Issues

- #56 Control Daemon for Crash Recovery (complementary - daemon can supervise background jobs)
- #21 Ralph Loop State Persistence (enables resume after job restart)

## Labels

enhancement

### PR DESCRIPTION
**Pull request by jpshackelford**
## Summary

Implements GitHub issue #68: Background job management for long-running commands.

## Features

- **Background execution**: Run commands with `--background` flag on `implement` and `refine`
- **Custom job names**: Use `--job-name` to set a custom name for the job
- **Detached processes**: Jobs survive terminal exit (uses `setsid`/`start_new_session`)
- **Job metadata**: Stored in `~/.lxa/jobs/` with automatic status updates on completion

## Job Control Commands

```bash
lxa job list                    # List all jobs with status
lxa job status <id>             # Show detailed job info
lxa job logs <id>               # View job output
lxa job logs <id> --follow      # Tail logs in real-time
lxa job stop <id>               # Gracefully stop running jobs
lxa job clean --older-than 7    # Clean up old job files
```

## Usage Examples

```bash
# Start implementation in background
lxa implement --loop --background
# Output: Started job implement-a3f2b1c, logs at ~/.lxa/jobs/implement-a3f2b1c.log

# With custom name
lxa implement --loop --background --job-name overnight-run

# Refine a PR in background
lxa refine https://github.com/owner/repo/pull/42 --background --job-name pr42-refine
```

### New Package: `src/jobs/`

| Module | Description |
|--------|-------------|
| `models.py` | `Job` dataclass and `JobStatus` enum |
| `storage.py` | JSON file persistence in `~/.lxa/jobs/` |
| `manager.py` | `JobManager` class for CRUD operations |
| `executor.py` | `spawn_detached()` for background process spawning |
| `cli/` | CLI presentation layer for job commands |

## Tests

75 new tests covering all functionality:
- `test_models.py`: 18 tests for Job and JobStatus
- `test_storage.py`: 19 tests for storage operations  
- `test_manager.py`: 25 tests for JobManager
- `test_executor.py`: 13 tests for detached execution

## Checklist

- [x] `--background` flag starts detached process that survives terminal exit
- [x] `--job-name` allows custom naming with auto-appended hash
- [x] `lxa job list` shows running and recent jobs
- [x] `lxa job status <id>` shows detailed job info
- [x] `lxa job logs <id>` displays log contents
- [x] `lxa job logs <id> --follow` tails logs in real-time
- [x] `lxa job stop <id>` gracefully terminates running jobs
- [x] Jobs directory created at `~/.lxa/jobs/`
- [x] Metadata persists across restarts

Closes #68

---
*This PR was created by an AI assistant (OpenHands) on behalf of jpshackelford.*

### CODE REVIEW

File: src/jobs/models.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+        Args:
+            command: Command arguments to run
+            cwd: Working directory
+            jobs_dir: Directory for job files (e.g., ~/.lxa/jobs)
+            job_name: Custom job name (default: derived from command)
+
+        Returns:
+            New Job instance with generated ID and log path
+        """
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/jobs/models.py:111]: Fixed in ddc42a2. Added `sanitize_job_name()` function that:
- Replaces `/`, `\`, and `..` with dashes
- Removes null bytes and colons
- Collapses multiple dashes and strips leading/trailing dashes
- Falls back to "job" if result is empty

Added 9 tests covering path traversal, slashes, null bytes, and edge cases.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: src/jobs/executor.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+            stderr=subprocess.STDOUT,
+            start_new_session=True,  # This is the key - creates new session
+            env={**os.environ},  # Inherit environment
+        )
+
+        job.pid = proc.pid
+        save_job(job, jobs_path)
+
+        return job
+
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/jobs/executor.py]: Fixed in ddc42a2. Parent now calls `log_file.close()` immediately after `Popen` succeeds. The child process inherits the file descriptor before the parent closes its copy, so no data is lost.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: src/jobs/executor.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    with open(metadata_path) as f:
+        data = json.load(f)
+
+    data["status"] = "done" if exit_code == 0 else "failed"
+    data["exit_code"] = exit_code
+    data["ended_at"] = datetime.now().isoformat()
+
+    with open(metadata_path, "w") as f:
+        json.dump(data, f, indent=2)
+
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/jobs/executor.py]: Acknowledged. Keeping the string-based wrapper for now as a pragmatic trade-off - it's self-contained and avoids needing to locate/import a separate module from the detached process. Will consider refactoring to a separate wrapper module if this becomes a maintenance burden.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: src/jobs/executor.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    # open for the detached child process
+    log_file = open(log_path, "w")  # noqa: SIM115
+    try:
+        # Use the Python interpreter to run our wrapper script
+        proc = subprocess.Popen(
+            [sys.executable, "-c", wrapper_script],
+            cwd=str(cwd),
+            stdin=subprocess.DEVNULL,
+            stdout=log_file,
+            stderr=subprocess.STDOUT,
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/jobs/executor.py:83]: Fixed in ddc42a2. Now checks `proc.poll()` immediately after `Popen` returns. If the process has already exited (poll returns non-None), we mark the job as FAILED with the exit code and return early.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: src/__main__.py [hunk 5]
```diff
@@ -933,6 +1131,43 @@ def main(argv: list[str] | None = None) -> int:
...
+        if args.auto_merge:
+            cmd.append("--auto-merge")
+        if args.allow_merge != "acceptable":
+            cmd.extend(["--allow-merge", args.allow_merge])
+        if args.min_iterations != 1:
+            cmd.extend(["--min-iterations", str(args.min_iterations)])
+        if args.max_refine_iterations != 5:
+            cmd.extend(["--max-refine-iterations", str(args.max_refine_iterations)])
+
+        job = spawn_lxa_command(
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/__main__.py:1169]: Fixed in ddc42a2 - same fix applied here. Both implement and refine now use the shared `_filter_background_args()` helper.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: src/jobs/manager.py [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        if job.pid is None:
+            return False, f"Job {job.id} has no PID"
+
+        try:
+            # Try graceful termination first
+            os.kill(job.pid, signal.SIGTERM)
+
+            # Wait for process to exit
+            import time
+
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/jobs/manager.py:174]: Acknowledged. This is a known limitation of PID-based process management. The race window is very small (milliseconds) and would require:
1. Our process to exit
2. OS to recycle the PID
3. New process to start with same PID
4. All before our SIGTERM is sent

For typical LXA usage (long-running jobs), this is extremely unlikely. If we need stronger guarantees in the future, we could track process start time or use process groups. Leaving as-is for now as a pragmatic trade-off.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: src/jobs/executor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/jobs/executor.py]: Actually, I went ahead and fixed this properly in 6843519.

Created `src/jobs/wrapper.py` as a proper module that can be invoked as:
```
python -m src.jobs.wrapper <job_id> <jobs_dir> -- <command...>
```

Benefits:
- Wrapper logic is now testable as normal Python code (11 new tests)
- No more repr(), json.dumps() embedding in strings
- Changes to wrapper logic are real Python edits, not string literal edits
- Executor simply builds a command list and calls Popen

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

**Comment by jpshackelford** [comment]
## Review Feedback Addressed ✅

All critical and important issues from the OpenHands review have been fixed:

### 🔴 Critical Issues - Fixed

| Issue | Commit | Fix |
|-------|--------|-----|
| **Path Traversal Risk** | ddc42a2 | Added `sanitize_job_name()` - removes `/`, `\`, `..`, null bytes, colons. 9 new tests. |
| **File Handle Leak** | ddc42a2 | Parent now calls `log_file.close()` after Popen succeeds. |

### 🟠 Important Issues - Fixed

| Issue | Commit | Fix |
|-------|--------|-----|
| **Fragile Code Generation** | 6843519 | Extracted to `src/jobs/wrapper.py` module. No more string-based code generation. 11 new tests. |
| **No Process Start Verification** | ddc42a2 | Added `proc.poll()` check after Popen to verify process started. |
| **Error-Prone Command Reconstruction** | ddc42a2 | Replaced manual flag checking with `_filter_background_args()` helper that filters argv. |

### 🟡 Acknowledged (Not Fixed)

| Issue | Rationale |
|-------|-----------|
| **Timing Attack Window** | Very rare edge case for long-running jobs. Would require process exit + PID recycle + new process start all within milliseconds. Pragmatic trade-off. |

### Test Coverage

- **95 total job tests** (up from 75 initially)
- All linting, formatting, and type checks pass

---
*This comment was posted by an AI assistant (OpenHands) on behalf of jpshackelford.*

File: uv.lock [hunk 7]
```diff
@@ -7,6 +7,30 @@ resolution-markers = [
     "python_full_version < '3.13'",
 ]
 
+[[package]]
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply uv.lock:10]: The lock file changes are from **transitive dependency updates** of existing packages, not from the jobs feature.

Looking at the actual diff:
- Only 7 package entries were added, 7 removed (net zero)
- The main change is `fastmcp` upgrading from 2.14.2 → 3.2.3
- The new packages (agent-client-protocol, aiofile, caio, etc.) are transitive deps of the upgraded fastmcp

The `src/jobs/` package itself uses **only stdlib imports** (os, signal, subprocess, json, datetime, pathlib). No new direct dependencies were added for this feature.

This happened because `uv sync` was run during development, which updated the lock file to the latest compatible versions. The timing is coincidental with this PR.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: src/jobs/manager.py [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+    Args:
+        job: Job to refresh
+        jobs_dir: Jobs directory for saving updates
+
+    Returns:
+        Updated job
+    """
+    if job.status != JobStatus.RUNNING:
+        return job
+
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/jobs/manager.py:69]: Good suggestion! Added this limitation to the docstring in ba8325c.

The docstring now explicitly documents that if the wrapper crashes after process success but before updating metadata (power loss, OOM kill, etc.), the job will incorrectly show as FAILED.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: src/jobs/executor.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        "--",
+        *command,
+    ]
+
+    # Start detached process
+    # - start_new_session=True creates a new session (like setsid)
+    # - stdin from /dev/null to fully detach
+    # - stdout/stderr to log file
+    # Note: We don't use context manager because we need to handle the file
+    # handle lifecycle carefully around Popen
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/jobs/executor.py:75]: Thanks for the positive feedback! The intentional non-context-manager pattern was a key design decision to ensure proper file descriptor inheritance to the child process.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: tests/jobs/test_executor.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+        assert job.id.startswith("echo-test-")
+        assert job.pid is not None
+        assert job.status == JobStatus.RUNNING
+
+    def test_process_survives(self, tmp_path: Path):
+        """Test spawned process runs independently."""
+        # Start a process that will take a moment
+        job = spawn_detached(
+            command=[sys.executable, "-c", "import time; time.sleep(2); print('done')"],
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply tests/jobs/test_executor.py:34]: Thanks! Testing with real processes was intentional - mocks would only verify "function was called" but wouldn't catch actual subprocess, signal handling, or file I/O issues. Real process tests catch real bugs.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._

File: src/__main__.py [hunk 11]
```diff
@@ -933,6 +1158,22 @@ def main(argv: list[str] | None = None) -> int:
...
 
+    # Handle background mode
+    if args.background:
+        from src.jobs import spawn_lxa_command
+
+        # Filter out --background and --job-name, keep everything else
+        args_to_use = argv if argv is not None else sys.argv[1:]
+        cmd = _filter_background_args(args_to_use)
+
+        job = spawn_lxa_command(
```
[the remark this answers is not in the document]
Comment (jpshackelford) [reply src/__main__.py:1169]: Agreed the current approach is acceptable. Using `argv or sys.argv[1:]` would incorrectly use `sys.argv[1:]` when `argv=[]` (empty list), which is a valid input. The explicit `is not None` check is more precise.

_This reply was posted by an AI assistant (OpenHands) on behalf of jpshackelford._
